modules/light-clients/08-wasm/types/validation_test.go

```diff
@@ -14,44 +22,141 @@ func TestValidateWasmCode(t *testing.T) {
 	testCases := []struct {
 		name     string
 		malleate func()
-		expPass  bool
+		expError error
 	}{
 		{
 			"success",
 			func() {
 				code, _ = os.ReadFile("../test_data/ics10_grandpa_cw.wasm.gz")
 			},
-			true,
+			nil,
 		},
 		{
-			"fails with empty byte slice",
+			"failure: empty byte slice",
 			func() {
 				code = []byte{}
 			},
-			false,
+			types.ErrWasmEmptyCode,
 		},
 		{
-			"fails with byte slice too large",
+			"failure: byte slice too large",
 			func() {
-				expLength := MaxWasmByteSize() + 1
+				expLength := types.MaxWasmByteSize() + 1
 				code = make([]byte, expLength)
 				length, err := rand.Read(code)
 				require.NoError(t, err, t.Name())
 				require.Equal(t, expLength, uint64(length), t.Name())
 			},
-			false,
+			types.ErrWasmCodeTooLarge,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc.malleate()
+
+		err := types.ValidateWasmCode(code)
+
+		if tc.expError == nil {
+			require.NoError(t, err, tc.name)
+		} else {
+			require.ErrorIs(t, err, tc.expError, tc.name)
+		}
+	}
+}
+
+func TestValidateWasmCodeHash(t *testing.T) {
+	var codeHash []byte
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success",
+			func() {
+				code, _ := os.ReadFile("../test_data/ics10_grandpa_cw.wasm.gz")
+				checksum := sha256.Sum256(code)
+				codeHash = checksum[:]
+			},
+			nil,
+		},
+		{
+			"failure: nil byte slice",
+			func() {
+				codeHash = nil
+			},
+			errorsmod.Wrap(types.ErrInvalidCodeHash, "code hash cannot be empty"),
+		},
+		{
+			"failure: empty byte slice",
+			func() {
+				codeHash = []byte{}
+			},
+			errorsmod.Wrap(types.ErrInvalidCodeHash, "code hash cannot be empty"),
+		},
+		{
+			"failure: byte slice size is not 32",
+			func() {
+				codeHash = []byte{1}
+			},
+			errorsmod.Wrapf(types.ErrInvalidCodeHash, "expected length of 32 bytes, got %d", 1),
+		},
+	}
+
+	for _, tc := range testCases {
+		tc.malleate()
+
+		err := types.ValidateWasmCodeHash(codeHash)
+
+		if tc.expError == nil {
+			require.NoError(t, err, tc.name)
+		} else {
+			require.ErrorContains(t, err, tc.expError.Error(), tc.name)
+		}
+	}
+}
+
+func TestValidateClientID(t *testing.T) {
+	var clientID string
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success: valid wasm client identifier",
+			func() {
+				clientID = defaultWasmClientID
```

Comment:
could also assign clientID before malleate and have empty func for success but not particularly bullish on that

Reply:
I don't think it is needed in this case as there are only 3 test cases. And it is good to see the clientID in each case since this is the only thing we are setting